File: Python/143. Reorder List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
                
        if not head:
            return 
        temp = head 
        slow = head 
        fast = head 
        
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        if fast:
            slow = slow.next 
        
        stack = [] 
        
        while slow:
            stack.append(slow)
            slow = slow.next 
        
        slow = head 
        
        while stack:
            temp = slow.next 
            node = stack.pop()
            slow.next = node 
            node.next = temp 
            slow = temp 
        
        slow.next = None 

File: 2024/143. Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL;
        while (head != NULL) {
            ListNode* next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
    void reorderList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return;
        }
        ListNode* slow = head, *fast = head->next;
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* rearHead = reverseList(slow->next);
        slow->next = NULL;
        ListNode* newHead = head;
        while (head != NULL && rearHead != NULL) {
            ListNode* headNext = head->next, *rearHeadNext = rearHead->next;
            head->next = rearHead;
            rearHead->next= headNext;
            head = headNext;
            rearHead = rearHeadNext;
        }
    }
};